src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -412,14 +413,41 @@ void LSHSearch<SortPolicy>::BuildHash()
   // Step III: Create each hash table in the first level hash one by one and
   // putting them directly into the 'secondHashTable' for memory efficiency.
   projections.clear(); // Reset projections vector.
+
+
+  if (projection.size() != 0 && projection.size() != numTables)
+    throw std::invalid_argument(
+        "number of projection tables provided must be equal to numProj"
+        );
+
   for (size_t i = 0; i < numTables; i++)
   {
     // Step IV: Obtain the 'numProj' projections for each table.
 
     // For L2 metric, 2-stable distributions are used, and
     // the normal Z ~ N(0, 1) is a 2-stable distribution.
     arma::mat projMat;
-    projMat.randn(referenceSet->n_rows, numProj);
+
+    if (projection.size() == 0) //random generation of table i
+    {
+
+      // For L2 metric, p-stable distributions are used, and the normal
+      // Z ~ N(0, 1) is p-stable.
+      projMat.randn(referenceSet->n_rows, numProj);
+    }
+    else //user-specified projection tables
+    {
+      projMat = projection[i];
+
+      //make sure specified matrix is of correct size
+      if ( projMat.n_rows != referenceSet->n_rows )
+        throw std::invalid_argument( 
+            "projection table dimensionality doesn't"
+            " equal dataset dimensionality" );
+      if ( projMat.n_cols != numProj )
+        throw std::invalid_argument(
+            "projection table doesn't have correct number of projections");
+    }
```

Comment:
We might be able to clean this up a bit: instead of calling `projMat.randn()`, if we use a `cube` instead, we can just call `cube::randn(referenceSet->n_rows, numProj, numTables)` once at the beginning of the function, and move this entire conditional there.  Then `projMat` or `projections[i]` won't need to be touched at all here.